### PR TITLE
fix: Support string type for the script field in Route

### LINKED ISSUE #1275: bug: the script field in Route is limited to map, instead of string
# Bug report

## Describe the bug

In APISIX, we can see the `script` field in `Route` is of type `string`.
https://github.com/apache/apisix/blob/3e19b06293463f52c8d26f0958542fc475ff0fcd/apisix/schema_def.lua#L456

While as creating a Route in manager-api, we forcibly cast the script to `map[string]interface{}`. 
https://github.com/apache/apisix-dashboard/blob/0f322e1d4a03e3416da1ae989fcfa49232035648/api/internal/handler/route/route.go#L332

So if sending a request to manager-api as below,

```
curl http://127.0.0.1:8000/apisix/admin/routes -X POST -H "Authorization: auth-header" -d '{"uri":"/hello","script":"local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\") \n end \nreturn _M","upstream":{"type":"roundrobin","nodes":{"127.0.0.1:1980":1}}}'
```

manager-api will always complain with the following error: 

```
err； interface conversion: interface {} is string, not map[string]interface {}
[Recovery] 2021/01/12 - 14:23:53 panic recovered:

interface conversion: interface {} is string, not map[string]interface {}
/usr/local/opt/go/libexec/src/runtime/iface.go:261 (0x100cd6e)
	panicdottypeE: panic(&TypeAssertionError{iface, have, want, ""})
/Users/imjoey/Work/apache-apisix/apisix-dashboard/api/internal/handler/route/route.go:332 (0x193d9fe)
```

As a contrast, the similar request can be successfully processed by APISIX Admin API.

Is such situation mentioned above as expected? Much appreciated. 

## Expected behavior

manager-api could also accept the `script` field of `string` type.

## System information

- OS: macOS
- Version: latest in master branch

## Additional 

This issue is also related to https://github.com/apache/apisix-dashboard/issues/1085, put here for tracking.

### PR DESCRIPTION
**Pull request by imjoey**
Signed-off-by: imjoey <majunjiev@gmail.com>

Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
Fixes #1275 .

___
### Bugfix
- Description

As discussed in #1275 , the `script` field in Route only support the configs generated by drag-to-lua, which actually acts not the same as what in APISIX Admin API. 

- How to fix?

Support the `script` field as a pure lua code.

### CODE REVIEW

File: api/internal/handler/route/route.go [hunk 1]
```diff
@@ -392,16 +398,14 @@ func (h *Handler) Update(c droplet.Context) (interface{}, error) {
...
-				fmt.Errorf("invalid `script`")
-		}
-		input.Route.Script, err = generateLuaCode(scriptConf)
-		if err != nil {
-			return &data.SpecCodeResponse{StatusCode: http.StatusInternalServerError}, err
+		if scriptConf, ok := input.Script.(map[string]interface{}); ok {
+			input.Route.Script, err = generateLuaCode(scriptConf)
+			if err != nil {
+				return &data.SpecCodeResponse{StatusCode: http.StatusInternalServerError}, err
+			}
```
Comment (starsz) [inline api/internal/handler/route/route.go:408]: Hi, @imjoey, we had `loadstring` in apisix to validate the script.
Can we do same things in dashboard?

Comment (imjoey) [reply api/internal/handler/route/route.go:425]: @starsz thanks for your advice. PR renewed and an additional validation check was added. Looking forward to your feedback. Much appreciated.

File: api/internal/utils/utils.go [hunk 2]
```diff
@@ -161,3 +162,10 @@ func LabelContains(labels map[string]string, reqLabels map[string]struct{}) bool
 
 	return false
 }
+
+// ValidateLuaCode validates lua syntax for input code, return nil
+// if passwd, otherwise a non-nil error will be returned
```
Comment (starsz) [inline api/internal/utils/utils.go:167]: passwd? Is this a typo?

File: api/internal/handler/route/route_test.go [hunk 3]
```diff
@@ -1349,3 +1349,146 @@ func Test_Route_With_Script(t *testing.T) {
...
+		  "upstream": {
+		      "type": "roundrobin",
+		      "nodes": [{
+		          "host": "www.a.com",
+		          "port": 80,
+		          "weight": 1
+		      }]
+		  },
+		  "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\") \n end \nreturn _M"
+		}`
```
Comment (starsz) [inline api/internal/handler/route/route_test.go:1383]: Style: JSON format looks not very good.
I can see different  JSON format in your code, maybe you should format them like

https://github.com/apache/apisix-dashboard/pull/1289/files#diff-69dbad33b4168d98e52aecaff8a2997d5101e9936a45adf502a3cde9cb9f1763R1406-R1418

File: api/internal/handler/route/route_test.go [hunk 4]
```diff
@@ -1349,3 +1349,146 @@ func Test_Route_With_Script(t *testing.T) {
...
+		  "upstream": {
+		      "type": "roundrobin",
+		      "nodes": [{
+		          "host": "www.a.com",
+		          "port": 80,
+		          "weight": 1
+		      }]
+		  },
+		  "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\")"
+		}`
```
Comment (starsz) [inline api/internal/handler/route/route_test.go:1488]: Here

File: api/test/e2e/route_with_script_luacode_test.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+					 "upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1980,
+							"weight": 1
+						}]
+					 },
+					 "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\") \n end \nreturn _M"
+				 }`,
```
Comment (starsz) [inline api/test/e2e/route_with_script_luacode_test.go:42]: Ditto.

File: api/test/e2e/route_with_script_luacode_test.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+					 "upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1980,
+							"weight": 1
+						}]
+					 },
+					 "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\")"
+				 }`,
```
Comment (starsz) [inline api/test/e2e/route_with_script_luacode_test.go:92]: Ditto.

File: api/test/e2e/route_with_script_luacode_test.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+					 "upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1981,
+							"weight": 1
+						}]
+					 },
+					 "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\") \n end \nreturn _M"
+				 }`,
```
Comment (starsz) [inline api/test/e2e/route_with_script_luacode_test.go:72]: Ditto.

File: api/test/e2e/route_with_script_luacode_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+					 "upstream": {
+						"type": "roundrobin",
+						"nodes": [{
+							"host": "172.16.238.20",
+							"port": 1980,
+							"weight": 1
+						}]
+					 },
+					 "script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\")"
+				 }`,
```
Comment (starsz) [inline api/test/e2e/route_with_script_luacode_test.go:121]: Ditto.

File: api/internal/handler/route/route.go (hunk 1 above)
Comment (starsz) [reply api/internal/handler/route/route.go:425]: Well done! Thanks a lot.

File: api/internal/utils/utils.go (hunk 2 above)
Comment (imjoey) [reply api/internal/utils/utils.go:167]: > passwd? Is this a typo?

@starsz oh god, yep, it's a typo. Good catch. Will fix it soon. Thanks.

File: api/internal/handler/route/route_test.go (hunk 3 above)
Comment (imjoey) [reply api/internal/handler/route/route_test.go]: @starsz PR renewed. The style will be fine if we only use  `table` to indent the JSON defined as a string type. Thanks for advice.

File: api/internal/handler/route/route_test.go (hunk 4 above)
Comment (imjoey) [reply api/internal/handler/route/route_test.go]: @starsz  Done. Thanks.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 5 above)
Comment (imjoey) [reply api/test/e2e/route_with_script_luacode_test.go]: @starsz  Done. Thanks.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 6 above)
Comment (imjoey) [reply api/test/e2e/route_with_script_luacode_test.go]: @starsz  Done. Thanks.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 8 above)
Comment (imjoey) [reply api/test/e2e/route_with_script_luacode_test.go]: @starsz  Done. Thanks.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 7 above)
Comment (imjoey) [reply api/test/e2e/route_with_script_luacode_test.go]: @starsz  Done. Thanks.

**Comment by starsz** [comment]
Thanks a lot.`Semantic Pull Request ` CI failed.
You need to fix the PR title.

**Comment by juzhiyuan** [comment]
> Thanks a lot.`Semantic Pull Request ` CI failed.
> You need to fix the PR title.

image

This is not a required check :) we could safely ignore it.

**Comment by juzhiyuan** [comment]
Please resolve related comments if needed.

**Comment by starsz** [comment]
> > Thanks a lot.`Semantic Pull Request ` CI failed.
> > You need to fix the PR title.
> 
> image
> 
> This is not a required check :) we could safely ignore it.

IMO, I think we need to fix it. In this case, the prefix of the title should be `fix`.

**Comment by imjoey** [comment]
@starsz @juzhiyuan All tests passed. Thank for suggestions.

**Review by starsz — APPROVED** [review]

**Comment by juzhiyuan** [comment]
ping @nic-chen @tokers

**Review by tokers — APPROVED** [review]

File: api/test/e2e/route_with_script_luacode_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+		{
+			Desc:         "get the route",
+			Object:       ManagerApiExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/apisix/admin/routes/r1",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			ExpectBody:   "hit access phase",
+			Sleep:        sleepTime,
+		},
```
Comment (nic-chen) [inline api/test/e2e/route_with_script_luacode_test.go:55]: it should be better if we could verify the `script` on DP.

File: api/internal/handler/route/route_test.go [hunk 10]
```diff
@@ -1349,3 +1349,146 @@ func Test_Route_With_Script(t *testing.T) {
...
+				"weight": 1
+			}]
+		},
+		"script": "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.INFO,\"hit access phase\")"
+	}`
+	err = json.Unmarshal([]byte(reqBody), route)
+	assert.Nil(t, err)
+	ctx.SetInput(route)
+	_, err = handler.Create(ctx)
+	assert.NotNil(t, err)
```
[the remark this answers is not in the document]
Comment (imjoey) [reply api/internal/handler/route/route_test.go:1493]: @nic-chen thanks for advice. PR renewed.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 9 above)
Comment (imjoey) [reply api/test/e2e/route_with_script_luacode_test.go:62]: @nic-chen PR renewed. Additional check for the log generated by `Script` is added. PTAL. Thanks

**Comment by imjoey** [comment]
@tokers  @nic-chent  @juzhiyuan @starsz 

**Help Wanted:**

In order to verify the `script` in APISIX, I read the access log of APISIX to check if the content has the string `hit access phase`. While the check result is always false.

Honestly I'm new to `script`, did I miss anything or misunderstand the usage? Looking forward to your helps. Much appreciated.

**Comment by nic-chen** [comment]
> @tokers @nic-chent @juzhiyuan @starsz
> 
> **Help Wanted:**
> 
> In order to verify the `script` in APISIX, I read the access log of APISIX to check if the content has the string `hit access phase`. While the check result is always false.
> 
> Honestly I'm new to `script`, did I miss anything or misunderstand the usage? Looking forward to your helps. Much appreciated.

@imjoey 
You could test locally to see if the log can be written successfully?

**Comment by imjoey** [comment]
@nic-chen

it's not in my local neither. I'll refer to OpenResty to look for the reason. Thanks.

**Comment by nic-chen** [comment]
> @nic-chen
> 
> it's not in my local neither. I'll refer to OpenResty to look for the reason. Thanks.

@imjoey  `ngx.log` will write a log in `error.log`, not `access.log`.

**Comment by imjoey** [comment]
> > @nic-chen
> 
> > 
> 
> > it's not in my local neither. I'll refer to OpenResty to look for the reason. Thanks.
> 
> 
> 
> @imjoey  `ngx.log` will write a log in `error.log`, not `access.log`.

@nic-chen  You are my life saver. Much appreciated. I'll check soon.

**Review by gxthrj — APPROVED** [review]
LGTM

**Comment by gxthrj** [comment]
But e2e test  failed.

File: api/test/e2e/route_with_script_luacode_test.go (hunk 9 above)
Comment (nic-chen) [reply api/test/e2e/route_with_script_luacode_test.go:62]: Thanks.

**Comment by imjoey** [comment]
> But e2e test failed.

@nic-chen  @gxthrj  CI passed, PTAL. Thanks.

**Pull request merged**